### PR TITLE
HMRC-2529: Migrate managed WAF rules to standalone aws_wafv2_web_acl_rule resources

### PR DESCRIPTION
**Pull request by AigulDj**
## What:

- Migrated WAF rules from inline rule blocks on `aws_wafv2_web_acl` to standalone `aws_wafv2_web_acl_rule` resources (managed AWS rule groups, IP set rules, IP+URL rate-based rules, filtered header rules, rule group references, URI path match rules, header allow rules, Bot Control), using the provider's create-or-adopt pattern so live rules are adopted by name, not recreated.
- Added lifecycle `{ ignore_changes = [rule] }` to `aws_wafv2_web_acl.this`, required whenever rules are managed via `aws_wafv2_web_acl_rule.`
- Restructured Bot Control's URI-prefix exclusion: added a separate allow rule at priority 69 (just ahead of Bot Control at 70) matching the excluded paths, replacing a `not_statement` that the standalone resource's schema doesn't support in this location.
- Left `ip_rate_based_rule` inline and frozen (documented in-code) - its exclusion needs the same restructure but has a bigger blast radius; flagged for a follow-up product decision.
- Fixed a dormant bug in `group_rules:` undefined `excluded_rule.value` reference, corrected to `rule_action_override.value.`
- Split `header_allow_rules` into non-sensitive metadata + a new sensitive `header_allow_values` map, since Terraform blocked using the combined list as a `for_each` source once it contained sensitive header values.
- Added `TF_VAR_WAF_E2E_SECRET_TOKEN` to plan-development's env, so its plan output matches what apply actually does.

## Why:

Inline rule blocks are a `TypeSet` with no stable per-rule diffing, so any WAF rule change produced a noisy, opaque plan diff on every unrelated PR (sensitive-value masking made it worse - dozens of unreadable ~/-/+ blocks). Standalone resources give each rule its own state entry, so changes are scoped to the rule that actually changed.

## Ticket:
[HMRC-2529](https://transformuk.atlassian.net/browse/HMRC-2529)

## Risk:
**Risk level:** 🟠

**Reason for rating:**

- Rule adoption depends on exact name/priority match to the live ACL; verified via terraform plan showing no diff for adopted rules before merge.
- Bot Control's restructure is a real behaviour change: excluded URI prefixes now skip Bot Control via an allow rule that terminates evaluation, so they'd also skip any future rule added with priority > 70. Documented in-code; matches the original intent for currently-existing rules.
- header_allow_rules/header_allow_values split touches root module inputs - downstream references were checked, but worth a second look in review.
- ip_rate_based_rule is unchanged in behaviour (still inline) but now frozen from Terraform's perspective; manual ignore_changes removal required for any future edit until migrated properly.
- No expected change to production WAF behaviour on apply - this is a Terraform management-layer migration, not a rules/policy change, aside from the Bot Control restructure noted above.

### CODE REVIEW

**Review by willfish — DISMISSED** [review]

File: environments/production/common/waf.tf [hunk 1]
```diff
@@ -24,24 +24,27 @@ module "waf" {
   }
 
   header_allow_rules = concat(
-    var.waf_mcp_secret_token != "" ? [
+    nonsensitive(var.waf_mcp_secret_token != "") ? [
```
Comment (amberstarlight) [inline environments/production/common/waf.tf:27]: Are we sure this is ok to be non-sensitive?

Comment (AigulDj) [reply environments/production/common/waf.tf:27]: Yep. The non-sensitive data only contains the rule configuration required to build the WAF rules. The secret header values are stored separately in `header_allow_values` and remain sensitive. Splitting them out also removes the Terraform WAF plan noise that was marking entire rule blocks as sensitive.
Screenshot 2026-07-23 at 11 49 32

File: modules/waf/main.tf [hunk 2]
```diff
@@ -174,392 +116,482 @@ resource "aws_wafv2_web_acl" "this" {
     }
   }
 
-  dynamic "rule" {
-    for_each = var.ip_rate_url_based_rules
-    content {
-      name     = rule.value.name
-      priority = rule.value.priority
+  lifecycle { ignore_changes = [rule] }
```
Comment (willfish) [inline modules/waf/main.tf:124]: This freezes `ip_rate_based_rule` too, so limit, action and telemetry changes will produce no diff. Can we migrate that rule before ignoring the whole rule set?

File: modules/waf/main.tf [hunk 3]
```diff
@@ -174,392 +116,482 @@ resource "aws_wafv2_web_acl" "this" {
...
+    metric_name                = each.value.name
+    sampled_requests_enabled   = true
+  }
+}
 
-    content {
-      name     = rule.value.name
-      priority = rule.value.priority
+resource "aws_wafv2_web_acl_rule" "allow_bot_control_excluded_paths" {
+  for_each = var.bot_control_rule != null ? {
```
Comment (willfish) [inline modules/waf/main.tf]: `excluded_uri_prefixes = []` is valid, but this still creates an empty `or_statement` and fails planning. Can we only create this rule when exclusions exist?

File: modules/waf/variables.tf [hunk 4]
```diff
@@ -177,16 +177,22 @@ variable "uri_path_match_rules" {
...
-    header_value = string
+    name        = string
+    priority    = number
+    header_name = string
   }))
   default = []
 }
 
+variable "header_allow_values" {
+  description = "Sensitive header values, keyed by rule name"
```
Comment (willfish) [inline modules/waf/variables.tf:217]: Can we validate these keys against `header_allow_rules[*].name`? Drift currently fails later with an opaque invalid-index error.

File: modules/waf/main.tf [hunk 5]
```diff
@@ -174,392 +116,482 @@ resource "aws_wafv2_web_acl" "this" {
...
+    "allow-bot-control-excluded-paths" = var.bot_control_rule
+  } : {}
 
-      action {
-        allow {}
-      }
+  web_acl_arn = aws_wafv2_web_acl.this.arn
+  name        = "allow-bot-control-excluded-paths"
+  priority    = 69 # must run immediately before bot_control (70)
```
Comment (willfish) [inline modules/waf/main.tf]: This needs to derive from `bot_control_rule.priority`; changing Bot Control from 70 otherwise breaks the ordering.

File: modules/waf/main.tf [hunk 6]
```diff
@@ -174,392 +116,482 @@ resource "aws_wafv2_web_acl" "this" {
...
+}
+
+resource "aws_wafv2_web_acl_association" "this" {
+  count = var.associate_alb ? 1 : 0
+
+  resource_arn = var.alb_arn
+  web_acl_arn  = aws_wafv2_web_acl.this.arn
+}
+
+resource "aws_wafv2_web_acl_rule" "managed" {
```
Comment (willfish) [inline modules/waf/main.tf]: `aws_wafv2_web_acl_rule` requires AWS provider 6.37+, but the module still declares `>= 6.0.0`.

File: modules/waf/main.tf [hunk 7]
```diff
@@ -174,392 +116,482 @@ resource "aws_wafv2_web_acl" "this" {
...
-              positional_constraint = "EXACTLY"
-              search_string         = rule.value.host
-
-              field_to_match {
-                single_header {
-                  name = "host"
-                }
-              }
+  action {
+    allow {}
```
Comment (willfish) [inline modules/waf/main.tf:615]: This `allow` stops ACL evaluation, so these paths bypass any future rules after priority 69. Can we keep the exclusion scoped to Bot Control?

File: modules/waf/main.tf [hunk 8]
```diff
@@ -1,10 +1,21 @@
 # https://docs.aws.amazon.com/waf/latest/developerguide/how-aws-waf-works.html
 
+locals {
+  filtered_header_rules = {
+    for header_name in var.filtered_header_rule.header_types :
+    replace(header_name, ".", "-") => {
+      priority     = var.filtered_header_rule.priority + index(var.filtered_header_rule.header_types, header_name) + 1
```
Comment (willfish) [inline modules/waf/main.tf]: Use `for i, header_name in ...` here. `index()` rescans the list and returns the first position for duplicates.

File: modules/waf/main.tf [hunk 9]
```diff
@@ -174,392 +116,482 @@ resource "aws_wafv2_web_acl" "this" {
...
+  count = var.associate_alb ? 1 : 0
+
+  resource_arn = var.alb_arn
+  web_acl_arn  = aws_wafv2_web_acl.this.arn
+}
+
+resource "aws_wafv2_web_acl_rule" "managed" {
+  for_each = var.managed_rules
+
+  web_acl_arn = aws_wafv2_web_acl.this.arn
```
Comment (willfish) [inline modules/waf/main.tf]: Each standalone rule reads the full ACL and updates are serialised. Have we accepted the extra plan/apply time as the rule count grows?

File: modules/waf/main.tf [hunk 10]
```diff
@@ -174,392 +116,482 @@ resource "aws_wafv2_web_acl" "this" {
...
-        sampled_requests_enabled   = true
-      }
+  visibility_config {
+    cloudwatch_metrics_enabled = true
+    metric_name                = each.key
+    sampled_requests_enabled   = true
+  }
+}
+
+resource "aws_wafv2_web_acl_rule" "ip_sets" {
```
Comment (willfish) [inline modules/waf/main.tf:162]: Can we add module tests for this migration? The environment plans don't exercise several of these rule families or Bot Control edge cases.

File: modules/waf/main.tf (hunk 2 above)
Comment (AigulDj) [reply modules/waf/main.tf:124]: Good catch, and worth being upfront about rather than hiding behind the lifecycle block.

Short answer 🙈
not yet, the standalone `aws_wafv2_web_acl_rule` resource doesn't support `not_statement/or_statement` inside `rate_based_statement.scope_down_statement` (confirmed against the live provider schema, not just docs), which is what this rule uses to exclude cached asset paths from rate limiting. 
The same workaround I used for Bot Control (a preceding allow rule) doesn't translate safely here: Bot Control sits at priority **70**, so the workaround only lets excluded paths skip Bot Control itself. This rule sits at priority **1**, so an equivalent allow rule would need to run before nearly everything else in the ACL, meaning asset paths would skip rate limiting and all managed rule groups, SQLi, Bot Control, etc., not just rate limiting. 
That's a real security-scope change, not a mechanical migration, so I didn't want to make that call inside this PR. I should probably raise it before merging the other changes

File: modules/waf/main.tf (hunk 6 above)
Comment (AigulDj) [reply modules/waf/main.tf]: Yep, when I checked provider version used it was showing 6.51, so my thought was the >= 6.0 constraint is enough, but maybe worth changing to >=  6.37.0

File: modules/waf/main.tf (hunk 9 above)
Comment (AigulDj) [reply modules/waf/main.tf]: That's a good point. I hadn't specifically measured the impact on plan/apply times. The main goal of this change was to remove the Terraform noise caused by rule updates and sensitive values propagating through the ACL. I think our rule count is fairly small, but it's worth considering whether the operational overhead becomes noticeable as more rules are added.

File: modules/waf/main.tf (hunk 7 above)
Comment (AigulDj) [reply modules/waf/main.tf:615]: Unfortunately not, because a standalone rule resource schema doesn't support `not_statement.` I added a code comment stating the invariant explicitly (any future rule must have priority < 69, or this exclusion silently swallows it) and added a test asserting Bot Control remains the highest-priority rule in the module's default config, that at least turns a silent bypass into a caught test failure if the invariant is ever broken.

File: modules/waf/main.tf (hunk 2 above)
Comment (AigulDj) [reply modules/waf/main.tf:124]: As discussed during the SU, we can migrate this once it's confirmed that it's safe for asset paths to bypass all existing WAF rules.

File: .github/workflows/deploy-to-development.yml [hunk 11]
```diff
@@ -75,8 +75,14 @@ jobs:
 
       - run: pre-commit run --all-files --show-diff-on-failure
 
+      - name: Run WAF module tests
+        run: terraform test
+        working-directory: modules/waf
```
Comment (willfish) [inline .github/workflows/deploy-to-development.yml]: Does this duplicate https://github.com/trade-tariff/trade-tariff-platform-aws-terraform/blob/a35152c4a95bea0e848e971b8fa95ec1da4f8508/.github/workflows/deploy-to-development.yml#L72?

**Review by willfish — DISMISSED** [review]

**Review by neilmiddleton — DISMISSED** [review]

File: .github/workflows/deploy-to-development.yml (hunk 11 above)
Comment (AigulDj) [reply .github/workflows/deploy-to-development.yml]: Good catch, seems like yes it is

**Review by neilmiddleton — APPROVED** [review]
